### ISSUE TITLE
Applying dead code detection in a Symfony Application

### ISSUE THREAD

**Issue by tomtomau** [vimeo/psalm#1105]
Hi @muglug - thanks so much for all your work on Psalm, it's an integral part of our workflow!

Context:
* Symfony 4 application
* lots of DI

After coming across a few places we had unused code, a colleague reminded me there is dead code detection available in Psalm. After enabling this flag, we were _blessed_ by 1400 errors (we otherwise have 0).

Categorising the errors:
* PossiblyUnusedMethod to `__construct` calls that are loaded with DI
* PossiblyUnusedMethod to functions that get executed with [event subscribers](https://symfony.com/doc/current/event_dispatcher.html#creating-an-event-subscriber)
* UnusedMethods and UnusedClasses to Doctrine Entity getters/setters that are only used with Symfony Forms, Controller methods and Command methods
* Genuine unused classes/methods

What I'm seeking is a bit of direction here, there's obviously a lot of Framework Magic™️ going on with Symfony that I see as well outside of Psalm's scope to interpret, but there's a lot of value available if it can be applied.

My thinking around an approach is to whitelist directories/namespaces that we care about running dead code check on, for us would be `/Services/`, `/Core/` (business logic core functions) and a selection of use case specific namespaces. This _should_ effectively limit the scope of Psalm to non-framework-magic code.

This _should_ leave genuine cases, plus the `__construct` calls. Referring to the docs to identify how to surpress errors, it would be at least possible to doc each `__construct` manually. In an ideal case, we could use the xml configuration to ignore a regex pattern matching `__construct` like:

```xml
  <PossiblyUnusedMethod>
    <errorLevel type="suppress">
      <referencedMethod pattern="__construct" />
    </errorLevel>
  </PossiblyUnusedMethod>
```

I think having that, it would be fairly trivial to apply Psalm for dead code detection on critical code (that is not tied to a framework). I plan to test this out and see if there's any other issues I run into but I would be keen to hear if you had any other ideas @muglug (and the community in general).

I would be happy to then contribute documentation about this as well.

**Comment by tomtomau** [comment]
That should - but I still have `__constructs` in those namespaces that I want to categorically ignore. The only option for now to ignore the constructs appears to be explicitly listing each `__construct`, either in `@annotations` or in the XML file.

**Comment by muglug** [comment]
I've added support for `*::methodName`, allowing you to do _just that thing_. If you want to add documentation, so much the better :)

**Comment by tomtomau** [comment]
Legend, thanks for such a quick turn around!!!
